### PR TITLE
Add more defines to be generated via Meson

### PR DESCRIPTION
**Pull request by dreamer**

### CODE REVIEW

**Review by kcgen — APPROVED** [review]
How much influence do we have over these define names? 

It would be nice to stick to a pattern, like
C_FEATURE
C_FUNCTION
HAVE_HEADER_H

For example: `#mesondefine C_HAVE_MPROTECT`

Is a function, but has HAVE in its name, when others don't, like `#mesondefine C_SET_PRIORITY`.

Tentatively approving assuming these defined variables cannot be changed by us (or would ripple changes through our sources and headers).

**Comment by dreamer** [comment]
I'll try to rename/fix defines that are used in one place only (on master) and then rebase this PR on top of new defines.

How about this naming scheme:

- `C_*` - project-related features and tweaks , e.g. `C_OPENGL` (I don't know why `C_` prefix was used in configure.ac; I suspect it was originally used for compiler feature checks and then misused for everything else; it's too late to change now).
- `HAVE_*_H` - header checks, e.g. `HAVE_STRINGS_H`
- `HAVE_*` - function/struct/struct field checks, so several defines wold be renamed, e.g.:
  -  `C_HAVE_MPROTECT` -> `HAVE_MPROTECT`
  - `DIRENT_HAS_D_TYPE` -> `HAVE_DIRENT_D_TYPE`
- Compiler feature checks: `C_HAS_ATTRIBUTE` (I have a proof of concept to remove this), `C_HAS_BUILTIN` (I think I'll manage to remove this define as well)

When implementing I decided to stick to existing define names, to avoid touching the code, but it's a good time to do some cleanup on master branch - some defines I already investigated:

- DB_HAVE_CLOCK_GETTIME - function check; function defined in POSIX; not available in Windows and MacOS X <= 10.12, seems to be available everywhere else (including Linux and FreeBSD) - used in 1 place
- HAVE_REALPATH - function check; function defined in POSIX; not available in Windows, but seems to be everywhere else (to be verified) - used in 1 place
- DIRENT_HAS_D_TYPE - struct field check; NON-POSIX (!); available in Linux and ported to (almost) all POSIX-compliant systems; not used on Windows IIRC (to double-check) - we use it in 1 place. This one is very important, it speeds up directory listing (considerably). I spent some time digging if there is any OS where it's not implemented, and the only one I found was Haiku.
- C_HAVE_MPROTECT - function check ; function defined in POSIX; this will be removed in W^X work, as it's not needed any more after W^X is implemented
- C_SET_PRIORITY - mainly function check; POSIX function ; used in 3 places in 1 file (but not in Windows; I would prefer to either remove it completely or maybe replace with SDL, not sure yet)

- WORDS_BIGENDIAN - Autoconf invention - need to stay as it is, used in too many places
- PAGESIZE - DOSBox invention; there are 2 or 3 defines for page size, this one means: pagesize of host/target CPU (not DOS pagesize); used extensively in CPU module, needs to stay as it is
- HAVE_NETINET_IN_H - header check - used in 1 place, maybe we can get rid of it, but not sure
- HAVE_PWD_H - header check - used in 1 place and perhaps can be removed once function will be moved to fs_utils.h / implemented separately for Windows and POSIX. Not sure yet.
- HAVE_STDLIB_H - header check - honestly I have no idea why this check exists and why anyone thought it's a good idea to use it. The only platform that I can think of, that might be missing it is MS-DOS.
- HAVE_STRINGS_H - header check - NON-POSIX (!) - `<strings.h>` header originates from BSD and was ported to Linux (looooong time ago). We use it and can't remove it ATM.
- HAVE_SYS_SOCKET_H - header check - NON-POSIX (!) - originates from BSD, available on Linux and probably all POSIX-compliant systems.
- HAVE_SYS_TYPES_H 1 - header check - POSIX. Even Windows provides a compatibility header, so we really don't need this one.
- ENVIRON_INCLUDED, ENVIRON_LINKED - not sure, I am trying to get rid of these two, but depending on what I try either macOS builds fail or Windows builds generate new warnings and I don't understand why; functionality itself is available everywhere. *quoting stackoverflow answers for Windows: "it's only a warning, don't worry about it" :/*
- HAVE_ALSA - This is really badly named… I will address in the future PR for Meson branch.

**Comment by kcgen** [comment]
All sounds good @dreamer; glad you can start steering these into a more consistent naming pattern while we have the opportunity.

I like the haves vs wants notation. 

C_HAS_ and HAVE_ are queried from the system and compiler, while the C_features (without HAS) are features we want to use.

File: meson.build [hunk 1]
```diff
@@ -71,21 +71,11 @@ if cc.has_member('struct dirent', 'd_type', prefix : '#include <dirent.h>')
...
-if cc.has_header('strings.h')
-  conf_data.set10('HAVE_STRINGS_H', true)
-endif
-
-if cc.has_header('sys/socket.h')
-  conf_data.set10('HAVE_SYS_SOCKET_H', true)
-endif
+foreach header : ['pwd.h', 'strings.h', 'netinet/in.h', 'sys/socket.h']
+  if cc.has_header(header)
+    conf_data.set('HAVE_' + header.underscorify().to_upper(), 1)
```
Comment (kcgen) [inline meson.build:76]: Beautiful :-)

Comment (dreamer) [reply meson.build:76]: :)

**Comment by dreamer** [comment]
For reference: #854

**Pull request merged**